userspace/libsinsp/container_engine/docker/podman.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright (C) 2021 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#include "podman.h"
+
+#include "container_engine/docker/lookup_request.h"
+#include "procfs_utils.h"
+#include "runc.h"
+#include "sinsp.h"
+
+#include <fstream>
+
+using namespace libsinsp::container_engine;
+using namespace libsinsp::runc;
+
+std::string podman::m_api_sock = "/run/podman/podman.sock";
+
+namespace {
+constexpr const cgroup_layout ROOT_PODMAN_CGROUP_LAYOUT[] = {
+	{"/libpod-", ".scope"}, // podman
+	{nullptr,    nullptr}
+};
+
+std::string get_systemd_cgroup(const sinsp_threadinfo *tinfo)
+{
+	// the kernel driver does not return cgroups without subsystems (e.g. name=systemd)
+	// in the cgroups field, so we have to do a check here, and load /proc/pid/cgroups
+	// ourselves if needed
+
+	for(const auto& it : tinfo->m_cgroups)
+	{
+		if(it.first == "name=systemd")
+		{
+			return it.second;
+		}
+	}
+
+	std::stringstream cgroups_file;
+	cgroups_file << scap_get_host_root() << "/proc/" << tinfo->m_tid << "/cgroup";
+
+	std::ifstream cgroups(cgroups_file.str());
+	return libsinsp::procfs_utils::get_systemd_cgroup(cgroups);
+}
+
+int get_userns_root_uid(const sinsp_threadinfo *tinfo)
+{
+	std::stringstream uid_map_file;
+	uid_map_file << scap_get_host_root() << "/proc/" << tinfo->m_tid << "/uid_map";
+
+	std::ifstream uid_map(uid_map_file.str());
+	return libsinsp::procfs_utils::get_userns_root_uid(uid_map);
+}
+
+// Check whether `tinfo` belongs to a podman container
+//
+// Returns the uid of the container owner:
+//  0 for root containers,
+//  >0 for rootless containers,
+//  NO_MATCH if the process is not in a podman container
+int detect_podman(const sinsp_threadinfo *tinfo, std::string& container_id)
+{
+	if(matches_runc_cgroups(tinfo, ROOT_PODMAN_CGROUP_LAYOUT, container_id))
+	{
+		return 0; // root
+	}
+
+	std::string systemd_cgroup = get_systemd_cgroup(tinfo);
+	if(systemd_cgroup.empty())
+	{
+		// can't get the cgroup name
+		return libsinsp::procfs_utils::NO_MATCH;
+	}
+
+	size_t pos = systemd_cgroup.find("podman-");
+	if(pos == std::string::npos)
+	{
+		return libsinsp::procfs_utils::NO_MATCH;
+	}
+
+	int podman_pid; // unused except to set the sscanf return value
```

Comment:
Unfortunately,

> The value EOF is returned if the end of input is reached before either the *first* successful conversion

So if just one field converts successfully,  we'll get a successful 0 return. I need the %c so that there is _something_ we can use as the container id.

I'm aware of `%*d` etc but IIRC I couldn't get the return value reliable enough to use for detecting the match/no match.

Reply:
You are right, and the laziness of saving 2 local vars would break everything. 
I feel stupid now! :D 
Thank you for the explanation.